app/uk/gov/hmrc/gform/LookupLoader.scala

```diff
@@ -51,7 +61,143 @@ class LookupLoader {
+      HeaderDecoder.decoder(englishLabel, welshLabel, id, keyWords, priority, countryCode)(
+        (
+          _: LookupLabel,
+          _: LookupLabel,
+          _: LookupId,
+          _: LookupKeyWords,
+          _: LookupPriority,
+          _: LookupCurrencyCountryCode))
+    val lookup = getClass.getClassLoader.getResourceAsStream("lookup/" + filename)
+    mkLookupType {
+      val cvsData
+        : List[(LookupLabel, LookupLabel, LookupId, LookupKeyWords, LookupPriority, LookupCurrencyCountryCode)] =
+        lookup
+          .unsafeReadCsv[
+            List,
+            (LookupLabel, LookupLabel, LookupId, LookupKeyWords, LookupPriority, LookupCurrencyCountryCode)](
+            rfc.withHeader)(headerDecoder, implicitly, implicitly)
+
+      val (enOptions, cyOptions): (List[(LookupLabel, LookupInfo)], List[(LookupLabel, LookupInfo)]) =
+        cvsData.zipWithIndex.map {
+          case ((enLabel, cyLabel, id, keyWords, priority, countryCode), idx) =>
+            val li = CurrencyLookupInfo(id, idx, keyWords, priority, countryCode)
+            ((enLabel, li), (cyLabel, li))
+        }.unzip
+
+      LocalisedLookupOptions {
+        Map(
+          LangADT.En -> LookupOptions(enOptions.toMap),
+          LangADT.Cy -> LookupOptions(cyOptions.toMap)
+        )
+      }
+    }
+  }
+
+  private def readPorts(
+    filename: String,
+    id: String,
+    englishLabel: String,
+    welshLabel: String,
+    keyWords: String,
+    priority: String,
+    region: String,
+    portType: String,
+    mkLookupType: LocalisedLookupOptions => LookupType): LookupType = {
+
+    val headerDecoder: HeaderDecoder[(
+      LookupLabel,
+      LookupLabel,
+      LookupId,
+      LookupKeyWords,
+      LookupPriority,
+      LookupRegion,
+      LookupPortType)] =
+      HeaderDecoder.decoder(englishLabel, welshLabel, id, keyWords, priority, region, portType)(
+        (
+          _: LookupLabel,
+          _: LookupLabel,
+          _: LookupId,
+          _: LookupKeyWords,
+          _: LookupPriority,
+          _: LookupRegion,
+          _: LookupPortType))
+    val lookup = getClass.getClassLoader.getResourceAsStream("lookup/" + filename)
+    mkLookupType {
+      val cvsData
+        : List[(LookupLabel, LookupLabel, LookupId, LookupKeyWords, LookupPriority, LookupRegion, LookupPortType)] =
+        lookup
+          .unsafeReadCsv[
+            List,
+            (LookupLabel, LookupLabel, LookupId, LookupKeyWords, LookupPriority, LookupRegion, LookupPortType)](
+            rfc.withHeader)(headerDecoder, implicitly, implicitly)
+
+      val (enOptions, cyOptions): (List[(LookupLabel, LookupInfo)], List[(LookupLabel, LookupInfo)]) =
+        cvsData.zipWithIndex.map {
+          case ((enLabel, cyLabel, id, keyWords, priority, region, portType), idx) =>
+            val li = PortLookupInfo(id, idx, keyWords, priority, region, portType)
```

Comment:
There is lots of duplication across `readPorts`, `readCurrencies`, `readCountries` and `read` methods.
For the start you can define local type alias:

```
type ColumnData =
      (LookupLabel, LookupLabel, LookupId, LookupKeyWords, LookupPriority, LookupRegion, LookupPortType)
```

Than reading from csv can be made generic for all methods to use:
```
private def readData[A](filename: String, headerDecoder: HeaderDecoder[A]): List[A] = {
    val lookup = getClass.getClassLoader.getResourceAsStream("lookup/" + filename)
    lookup.unsafeReadCsv[List, A](rfc.withHeader)(headerDecoder, implicitly, implicitly)
}
```

After that what will remain is to figure out how to generalise (there exists ways):

```
val (enOptions, cyOptions): (List[(LookupLabel, LookupInfo)], List[(LookupLabel, LookupInfo)]) =
        cvsData.zipWithIndex.map {
          case ((enLabel, cyLabel, id, keyWords, priority, region, portType), idx) =>
            val li = PortLookupInfo(id, idx, keyWords, priority, region, portType)
            ((enLabel, li), (cyLabel, li))
        }.unzip

      LocalisedLookupOptions {
        Map(
          LangADT.En -> LookupOptions(enOptions.toMap),
          LangADT.Cy -> LookupOptions(cyOptions.toMap)
        )
      }
```

Reply:
As @josef-vlach mentioned, boiler plate could be avoided. Possibly using a Type class to encode the specific behaviours

I have done a sample implementation here. Please take look

https://github.com/hmrc/gform-frontend/pull/969